### PR TITLE
[luci] Canonicalize user-given input/output_type

### PR DESCRIPTION
**Pull request by jinevening**
This parses input/output_type and canonicalize them.

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: https://github.com/Samsung/ONE/issues/10493
Draft PR: https://github.com/Samsung/ONE/pull/10518

### CODE REVIEW

File: compiler/luci/pass/src/CircleQuantizer.cpp [hunk 1]
```diff
@@ -268,6 +412,9 @@ void CircleQuantizer::quantize(loco::Graph *g) const
     if (output_type.empty())
       output_type = output_model_dtype;
 
+    auto input_type_vec = parse_comma_separated_str(input_type);
+    auto output_type_vec = parse_comma_separated_str(output_type);
```
Comment (jinevening) [inline compiler/luci/pass/src/CircleQuantizer.cpp:416]: Now user input is given as a comma-separated string.

File: compiler/luci/pass/src/CircleQuantizer.cpp [hunk 2]
```diff
@@ -49,6 +49,148 @@ namespace
 using namespace luci;
 using LayerParam = luci::CircleQuantizer::Options::LayerParam;
 
+// This function updates user-given input_type to match with the input signature of graph
+// If user gives only one input_type, it will be expanded to the number of graph inputs
```
Comment (jinevening) [inline compiler/luci/pass/src/CircleQuantizer.cpp:53]: This explains why canonicalization is introduced.

Previously, users only gave one input/output_type. For backward-compatibility, we still allow users to give one input/output_type. For models with multiple inputs/outputs, we expand the one input/output_type to match with the number of inputs/outputs of a graph. I'd call the process canonicalization.

File: compiler/luci/pass/src/CircleQuantizer.cpp [hunk 3]
```diff
@@ -323,8 +483,9 @@ void CircleQuantizer::quantize(loco::Graph *g) const
       ctx->input_model_dtype = str_to_dtype(input_model_dtype);
       ctx->output_model_dtype = str_to_dtype(output_model_dtype);
       ctx->granularity = str_to_granularity(granularity);
-      ctx->input_type = str_to_dtype(input_type);
-      ctx->output_type = str_to_dtype(output_type);
+      // TODO Change input/output_type to vector
```
Comment (jinevening) [inline compiler/luci/pass/src/CircleQuantizer.cpp:486]: This TODO will be resolved in the following PR.

File: compiler/luci/pass/src/helpers/Strings.h [hunk 4]
```diff
@@ -36,6 +36,10 @@ std::string to_lower_case(std::string);
 
 loco::DataType str_to_dtype(const std::string &);
 
+std::vector<std::string> parse_comma_separated_str(const std::string &);
```
Comment (mhs4670go) [inline compiler/luci/pass/src/helpers/Strings.h:39]: We could use existing library.

https://github.com/Samsung/ONE/blob/9ff5081c91f4fd9155f38beda649d0424a8b5a7a/compiler/pepper-csv2vec/include/pepper/csv2vec.h#L26

Comment (jinevening) [reply compiler/luci/pass/src/helpers/Strings.h:39]: Thanks for notice. I'll update source code to reduce duplicate code.

I'm a bit surprised that there is an individual module just for this purpose.

File: compiler/luci/pass/src/CircleQuantizer.cpp [hunk 5]
```diff
@@ -49,6 +50,148 @@ namespace
 using namespace luci;
 using LayerParam = luci::CircleQuantizer::Options::LayerParam;
 
+// This function updates user-given input_type to match with the input signature of graph
+// If user gives only one input_type, it will be expanded to the number of graph inputs
+void canonicalize_input_type(loco::Graph *g, std::vector<loco::DataType> &input_type)
+{
+  assert(g); // FIX_CALLER_UNLESS
```
Comment (mhs4670go) [inline compiler/luci/pass/src/CircleQuantizer.cpp:57]: Sometimes our static analysis tool complains about this assert on release build. Maybe we can do nothing for this case.

```suggestion
  if (g == nullptr)
    return;
    
  // ..
  
  if (inputs == nullptr)
    return;
```

File: compiler/luci/pass/src/CircleQuantizer.cpp [hunk 6]
```diff
@@ -49,6 +50,148 @@ namespace
...
+          input_type[i] != loco::DataType::S16)
+      {
+        throw std::runtime_error("Input dtype of " + input->name() +
+                                 " is invalid. Model's input is not float32.");
+      }
+    }
+  }
+}
+
+// This function updates user-given input_type to match with the input signature of graph
```
Comment (mhs4670go) [inline compiler/luci/pass/src/CircleQuantizer.cpp:122]: Maybe `input_type` -> `output_type`?

File: compiler/luci/pass/src/CircleQuantizer.cpp [hunk 7]
```diff
@@ -49,6 +50,148 @@ namespace
...
+          " is invalid. It has to be the same with the model's input dtype.");
+    }
+    else
+    {
+      // C2
+      if (input_type[i] != loco::DataType::FLOAT32 and input_type[i] != loco::DataType::U8 and
+          input_type[i] != loco::DataType::S16)
+      {
+        throw std::runtime_error("Input dtype of " + input->name() +
+                                 " is invalid. Model's input is not float32.");
```
[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/pass/src/CircleQuantizer.cpp:116]: I think current message is OK.

[the remark this answers is not in the document]
Comment (mhs4670go) [reply compiler/luci/pass/src/CircleQuantizer.cpp:116]: IIUC, 
`Model's input is not float32.` -> `Model's input is float32.`. Am I wrong? So, "Input dtype must be one of uint8, int16, float32" looks okay.

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/pass/src/CircleQuantizer.cpp:116]: Um...
Here, Model's input is float32. Message `Model's input is not float32.` is somewhat confusing, `input` can be understood to (1) Model or (2) User given.

IIUC, message should be `Input must be one of uint8, int16, float32` without `Model's`

[the remark this answers is not in the document]
Comment (jinevening) [reply compiler/luci/pass/src/CircleQuantizer.cpp:116]: @mhs4670go  @seanshpark  Thanks for finding. The error message looks confusing.

The error message is printed when user-given input is not one of {uint8 or int16 or float32} when model's input is float32.

I'll update the error message as below.

`For float32 input, the input type after quantization must be one of uint8, int16, or float32.`

**Comment by jinevening** [comment]
Ping @mhs4670go @seanshpark

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]
LGTM

**Pull request merged**